build(aio): big move of docs related files

petebacondarwin opened the pull request:
All the docs related files (docs-app, doc-gen, content, etc)
are now to be found inside the `/aio` folder.

The related gulp tasks have been moved from the top level
gulp file to a new one inside the `/aio` folder.

The structure of the `/aio` folder now looks like:

```
/aio/
  content/       #MARKDOWN FILES for devguides, cheatsheet, etc
    devguides/
    cheatsheets/
  transforms/    #dgeni packages, templates, etc
  webapp/           # the Angular app
    src/
      app/
      assets/
      content/  #HTML + JSON build artifacts produced by dgeni from /aio/content.
                #This dir is .gitignored-ed
```

petebacondarwin commented:
* TODO: remove node dependencies from angular/angular that are no longer required

petebacondarwin commented:
* TODO: I feel bad that there is a load of docs related build and test stuff spread throughout the ci-lite folder. I think it would be better if we could move all this to the `/aio/gulpfile.js`.

wardbell commented:
I was thinking that aio should have its own gulp folder. Undoubtedly there will need to be some way to kick off an aio gulp task or two at the top level. But it would be nice to treat `aio` as its own world, as if it were its own repo.

petebacondarwin commented:
@wardbell - that is mostly where this PR puts us. But I didn't want to do too much refactoring in such a monster of a commit.

petebacondarwin commented on .gitignore:
```diff
@@ -3,7 +3,6 @@
 /dist/
 node_modules
 bower_components
-angular.io/dist
```
because it is covered further down?

IgorMinar replied:
I believe that @petebacondarwin is right

gkalpak replied:
I can't see it. I agree it existed twice before, but you seem to have removed both occurences.

petebacondarwin replied:
sorry I just realised (I think) that the `ng build` process will compile to the dist folder; and that is where you are deploying to firebase from... I'll put that gitignore back in.

In the meantime it also means that I am not sure that `ng build` will be copying the generated docs to that folder....

petebacondarwin commented on scripts/ci-lite/test_aio.sh:
```diff
@@ -12,16 +12,13 @@ echo 'travis_fold:start:test.aio'
-
 # Lint the code
-cd "`dirname $0`/../../angular.io"
+cd "`dirname $0`/../../aio"
```
I don't think so because the yarn is running off the `package.json`, which I moved to the `/aio` folder

petebacondarwin replied:
But I think I need to change the `package.json`...

gkalpak replied:
Currently, there are files (e.g. `aio/webapp/src/tsconfig.json`,  `aio/webapp/e2e/tsconfig.json`,  `aio/webapp/src/angular-cli.json`) which assume (afaict) that the project root (i.e. the `node_modules/` directory, package.json etc) will be `aio/webapp/`. If we are moving the project root to `aio/`, when these files (and probably others) need to be updated accordingly.

petebacondarwin replied:
I just found out that it is not possible to specify the working directory (cwd) for the firebase command line tool. It must be run from the correct folder :-(
So the options are:

a) move the package.json back inside the webapp folder, next to the firebase.json file; which would mean that the stuff outside that folder does not benefit from the yarn install.
b) move the firebase.json out of the webapp folder; I think this might be the best option
c) change the package.json scripts to cd into that folder before running the firebase tool; I think this is not nice and probably would break on Windows

IgorMinar commented on aio/gulpfile.js:
```diff
@@ -0,0 +1,32 @@
+// THIS CHECK SHOULD BE THE FIRST THING IN THIS FILE
+// This is to ensure that we catch env issues before we error while requiring other dependencies.
+require('../tools/check-environment')({
+  requiredNpmVersion: '>=3.10.7 <4.0.0',
```
we don't use npm in aio

IgorMinar commented on aio/gulpfile.js:
```diff
@@ -0,0 +1,32 @@
+// This is to ensure that we catch env issues before we error while requiring other dependencies.
+require('../tools/check-environment')({
+  requiredNpmVersion: '>=3.10.7 <4.0.0',
+  requiredNodeVersion: '>=6.9.5 <7.0.0',
```
I wish we didn't duplicate this in /aio because now we have a yet another place to update the node version in this repo. could we reuse the top level check?

IgorMinar commented on aio/jsconfig.json:
```diff
@@ -0,0 +1,7 @@
+{
```
what's the purpose of this file? IDE support?

IgorMinar replied:
could we document it somehow? e.g. using `"comment": "the purpose of this file is... "`

IgorMinar reviewed (approved):
LGTM once we resolve the small things I pointed out.

petebacondarwin committed:
build(aio): big move of docs related files

All the docs related files (docs-app, doc-gen, content, etc)
are now to be found inside the `/aio` folder.

The related gulp tasks have been moved from the top level
gulp file to a new one inside the `/aio` folder.

The structure of the `/aio` folder now looks like:

```
/aio/
  build/         # gulp tasks
  content/       #MARKDOWN FILES for devguides, cheatsheet, etc
    devguides/
    cheatsheets/
  transforms/    #dgeni packages, templates, etc
  src/
    app/
    assets/
    content/    #HTML + JSON build artifacts produced by dgeni from /aio/content.
                #This dir is .gitignored-ed
  e2e/           #protractor tests for the doc viewer app
  node_modules/ #dependencies for both the doc viewer builds and the dgeni stuff
                #This dir is .gitignored-ed
  gulpfile.js   #Tasks for generating docs and building & deploying the doc viewer
```

Closes #14361
